### PR TITLE
Tapsigner

### PR DESCRIPTION
**Pull request by scgbckbone**
* add Tapsigner to HWI
* add Tapsigner simulator 
* testing
* adjust README.md in tests

### CODE REVIEW

**Comment by prusnak** [comment]
Please rebase on top of current master to remove the merge commit.

**Comment by scgbckbone** [comment]
> Please rebase on top of current master to remove the merge commit.

will do squash and rebase after this is reviewed and ready to go - I assume this will need some more commits (build is failing + implementation of review comments)

**Comment by scgbckbone** [comment]
Few notes here. We use password field for CVC, which is something like pin code but you have to provide it for every auth command. Therefore it is not suitable for us to use PIN (with sendpin/promptpin) as PIN does not unlock the card for some amount of time as with Coldcard or Trezor, instead we would need to use sendpin before every command which is not very practical to do.

* added togglepassphrase command which is interactive way how to change CVC. (keep in mind passphrase is CVC for Tapsigner and there is no way how to add bip39 passphrase to the card so I think it is ok)
* added restore command which will not restore the card as it is not possible, but is the way how to decrypt backup file with card a dump xprv to console (unsafe but I have a warning there)
* I have tested this with core v23.0 with all supported script types, you can check [here](https://blockstream.info/testnet/tx/02500d17a97c25a0153d457319ab2b676f4641e7529ce150cede90e4f6adea20) and [here](https://blockstream.info/testnet/tx/90bb23cb2420aae15f66f9c4f979207b394bfff8466cd8a59fe8b462ad1b835d) (signer is tapsigner simulator)
* with core you need to specify signer with CVC(password) on command line like this `-signer='./hwi.py -p 123456'` which definitely has some security implications for the user (but you still need to tap with the physical card).
* I have squashed and rebased on master as @prusnak proposed

Any review is greatly appreciated.

**Comment by achow101** [comment]
Please rebase now that CI is fixed.

**Comment by scgbckbone** [comment]
rebased on current master

File: hwilib/devices/tapsigner.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+    # In password variable we store card CVC
+    def __init__(self, path: str, password: str = "", expert: bool = False, chain: Chain = Chain.MAIN) -> None:
+        super(TapsignerClient, self).__init__(path, password, expert, chain)
+        self.device = find_card_by_path(path=path)
+
+    def interactive_get_cvc(self) -> None:
+        if not self.password:
+            cvc = getpass.getpass("Please provide CVC:\n")
+            if cvc:
+                self.password = cvc
```
Comment (achow101) [inline hwilib/devices/tapsigner.py:83]: We prefer to validate that the passphrase was provided in `__init__` rather than during the individual functions and interactively retrieving it. Having this optional interactivity breaks our model of non-interactivity for most functions.

File: hwilib/devices/tapsigner.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+        if not non_hardened:
+            self.device.set_derivation(path=bip32_path, cvc=self.password)
+            xp = self.device.get_xpub(self.password)
+            return ExtendedKey.deserialize(xp)
+        else:
+            str_path = path2str(hardened)
+            self.device.set_derivation(path=str_path, cvc=self.password)
+            xp = self.device.get_xpub(self.password)
+            ext_key = ExtendedKey.deserialize(xp)
+            return ext_key.derive_pub_path(non_hardened)
```
Comment (achow101) [inline hwilib/devices/tapsigner.py:119]: Does the Tapsigner only do hardened derivation? I don't generally like to have HWI do path validation and would much rather the device do it and let it give us an error.

File: hwilib/devices/tapsigner.py [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+            continue
+        if not card.is_tapsigner:
+            continue
+        d_data: Dict[str, Any] = {}
+        d_data['path'] = card.card_ident
+        d_data['type'] = 'tapsigner'
+        d_data['model'] = 'tapsigner'
+        d_data['label'] = None
+        d_data['needs_pin_sent'] = False
+        d_data['needs_passphrase_sent'] = False
```
Comment (achow101) [inline hwilib/devices/tapsigner.py:459]: Since `-p` must be specified, this should be `True`.

File: hwilib/devices/tapsigner.py [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+
+        if isinstance(card.tr, CKTapUnixTransport):
+            d_data['model'] += '_simulator'
+        if password:
+            # this is kind of unfortunate that we need password (a.k.a CVC) to get master fp
+            # because when using it with core, one has to provide password on the command line
+            # -signer='/home/more/PycharmProjects/HWI/venv/lib/python3.8/site-packages/hwi.py -p 123456'
+            d_data['fingerprint'] = card.get_xfp(password).hex().lower()
+        else:
+            d_data['fingerprint'] = None
```
Comment (achow101) [inline hwilib/devices/tapsigner.py:469]: If the password is not provided, this should set an `error` and `code` and omit `fingerprint`, as we do for the digitalbitbox and trezors which have a similar issue.

File: hwilib/devices/tapsigner.py (hunk 1 above)
Comment (scgbckbone) [reply hwilib/devices/tapsigner.py]: check for password in init - raise if not provided; remove interactive_get_cvc

File: hwilib/devices/tapsigner.py (hunk 2 above)
Comment (scgbckbone) [reply hwilib/devices/tapsigner.py]: yes, tapsigner only does hardened derivation... I moved the code to `coinkite-tap-protocol` https://github.com/coinkite/coinkite-tap-proto/pull/24 - after it is merged I will update `pyproject.toml` and `setup.py` with current version (for now there is a workaround)

**Comment by scgbckbone** [comment]
https://github.com/coinkite/coinkite-tap-proto/pull/24 merged

**Comment by scgbckbone** [comment]
updated `poetry.lock` with  `poetry update && poetry lock` after addition of `coinkite-tap-protocol`

File: hwilib/devices/tapsigner.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        fingerprint = hash160(extended_key_m.pubkey)[:4]
+        return fingerprint
+
+    def get_pubkey_at_path(self, bip32_path: str) -> ExtendedKey:
+        """
+        Get the public key at the BIP 32 derivation path.
+
+        :param bip32_path: The BIP 32 derivation path
+        :return: The extended public key
+        """
```
Comment (achow101) [inline hwilib/devices/tapsigner.py:73]: Docs don't need to be duplicated.

File: hwilib/devices/tapsigner.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        xpub = self.device.derive_xpub_at_path(self.password, fullpath=bip32_path)
+        return ExtendedKey.deserialize(xpub)
+
+    def sign_tx(self, tx: PSBT) -> PSBT:
+        # completely stolen and adjusted from digitalbox (blind signing)
```
Comment (achow101) [inline hwilib/devices/tapsigner.py:78]: Would it be possible for you to refactor this so that the sighashing code is shared rather than copy-pasting.

File: hwilib/devices/tapsigner.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+                    # Create tuples and add to List
+                    tup = (sighash, keypath.path, i_num, pubkey)
+                    sighash_tuples.append(tup)
+
+        # Return early if nothing to do
+        if len(sighash_tuples) == 0:
+            return tx
+
+        for sighash, int_path, i_num, pubkey in sighash_tuples:
+            rec_sig = self.device.sign_digest(cvc=self.password, slot=0, digest=sighash, fullpath=path2str(int_path))
```
Comment (achow101) [inline hwilib/devices/tapsigner.py:184]: What kind of exceptions will this throw? We need to be able to handle device exceptions and convert them into HWI exceptions so that the correct error codes come out. Usually this is done through decorators, see the other devices for an example.

**Comment by achow101** [comment]
Also, rebase for CI to be not as dead.

File: hwilib/devices/tapsigner.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+        if not password:
+            raise NoPasswordError('Password (cvc) must be supplied for Tapsigner')
+
+    def get_master_xpub(self, addrtype: AddressType = AddressType.WIT, account: int = 0) -> ExtendedKey:
+        bip44_pupose = get_bip44_purpose(addrtype)
+        bip44_chain = get_bip44_chain(self.chain)
+        path = f"m/{bip44_pupose}h/{bip44_chain}h/{account}h"
+        return self.get_pubkey_at_path(path)
+
+    def get_master_fingerprint(self) -> bytes:
```
[the remark this answers is not in the document]
Comment (scgbckbone) [reply hwilib/devices/tapsigner.py:59]: this has to stay, otherwise `0xffffffff`

File: hwilib/devices/tapsigner.py (hunk 6 above)
Comment (scgbckbone) [reply hwilib/devices/tapsigner.py]: added `get_sighash_tuples` as a member of `PSBT` class

File: hwilib/devices/tapsigner.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (scgbckbone) [reply hwilib/devices/tapsigner.py:59]: limitation of the Coinkite cards: XFP's are not calculated during derivation. `0xffffffff` as a placeholder

**Comment by scgbckbone** [comment]
not sure about CI --> seems to be full of random failures

File: hwilib/devices/tapsigner.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (scgbckbone) [reply hwilib/devices/tapsigner.py:59]: real card placeholder is `0x00000000` simulator had `0xffffffff` --> now fixed and both have real thing and simulator has `0x00000000`

**Comment by achow101** [comment]
Please squash/adjust so that each commit passes. Refactors should be in their own commit at the beginning.

**Comment by achow101** [comment]
I'm a bit concerned about the extra dependencies that this introduces. Up to know, HWI has only interacted with devices over USB, and the USB dependency stack is relatively well understood and under control. However this now introduces a completely new dependency stack to deal with NFC readers. Does this require us to ship new driver dependencies such as libusb? I'm highly hesitant to introduce a new device that is the only one that is going to use new driver dependencies.

**Comment by scgbckbone** [comment]
[52aa685f95b01a9480c0fb07acb7e433dc98ef2a](https://github.com/bitcoin-core/HWI/commit/52aa685f95b01a9480c0fb07acb7e433dc98ef2a) Moved Tapsigner and its dependencies to extras. This aims to partly alleviate your concerns with regards to `sdist` and `wheel` distributions. In sdist and wheel, only references to dependencies are stored. With normal `pip install ...` it won't install Tapsigner and its dependencies - so user has to specify `.[tapsigner]` or `-E tpsigner` to actually install/use it.

In this revision, however I still kept Tapsigner in binary distributions as I was not sure how "bit concerned" you are. I can remove Tapsigner from `build_{bin,wine}.sh` meaning that it would only ship by wheel and sdist not binaries. Or make separate binaries with tapsigner support? not sure what is your opinion on this @achow101 ?

**Comment by scgbckbone** [comment]
> [52aa685f95b01a9480c0fb07acb7e433dc98ef2a](https://github.com/bitcoin-core/HWI/commit/52aa685f95b01a9480c0fb07acb7e433dc98ef2a) Moved Tapsigner and its dependencies to extras. This aims to partly alleviate your concerns with regards to `sdist` and `wheel` distributions. In sdist and wheel, only references to dependencies are stored. With normal `pip install ...` it won't install Tapsigner and its dependencies - so user has to specify `.[tapsigner]` or `-E tpsigner` to actually install/use it.
> 
> In this revision, however I still kept Tapsigner in binary distributions as I was not sure how "bit concerned" you are. I can remove Tapsigner from `build_{bin,wine}.sh` meaning that it would only ship by wheel and sdist not binaries. Or make separate binaries with tapsigner support? not sure what is your opinion on this @achow101 ?

@prusnak @Sjors any opinions on this ?

**Comment by Sjors** [comment]
I don't understand Python dependencies that well. But if a lot a extra stuff is needed for NFS, can we (initially) have the user install those seperately and not fail if they're missing?

**Comment by scgbckbone** [comment]
> I don't understand Python dependencies that well. But if a lot a extra stuff is needed for NFC, can we (initially) have the user install those separately and not fail if they're missing?

This is already done with regards to python dependencies (tapsigner dependencies are part of extras). Namely `coinkite-tap-protocol` and `pyscard`. Not installed by default.

My question is whether you would accept additional dependencies (`swig` `libpcsclite-dev`) in binaries. Currently these two additional dependencies are needed for tapsigner to work. Other option is to not provide tapsigner support in binaries at all. User would need to install `swig` and `libpcsclite-dev` on their own and can only use `sdist` and `wheel` distributions with tapsigner explicitly specified as extra during installation (see above)

**Comment by scgbckbone** [comment]
Fair points. I will remove NFC dependencies from binaries and therefore `tapsigner` support will be purely optional and only available via `sdist` and `wheel`. Need to make tests work somehow. Thanks for your input @Sjors

**Comment by scgbckbone** [comment]
@Sjors I use HID Omnikey 5022 CL (if you want to play with coinkite NFC cards on desktop)

**Comment by Sjors** [comment]
I ordered the ACR122U because [a guy Reddit said so](https://www.reddit.com/r/linuxquestions/comments/kgil9c/comment/gggg5th/?utm_source=reddit&utm_medium=web2x&context=3).

**Comment by scgbckbone** [comment]
his reditt post is for `libnfc` which isn't what we use to talk to the reader (`sclite`) ... so doesn't apply

Seems like the one you choose coinkite lists as "not recommended" here https://github.com/coinkite/coinkite-tap-proto/blob/master/README.md#requirements 

Check coinkite store https://store.coinkite.com/store/category/accessories correct readers are available and there is also 25% discount today

**Comment by Sjors** [comment]
Oh man, why are there multiple NFC libraries for Linux?

**Comment by prusnak** [comment]
> Check coinkite store

"IMPORTANT: Incompatible with Mk4 NFC features! Works only with NFC cards."

Wow, so Coldcard and Tapsigner use incompatible NFC standards?

**Comment by scgbckbone** [comment]
Tapsigner is now completely optional and not included in binary distributions. Tests are handled to not fail if extras are not installed.

**Comment by achow101** [comment]
I've spent a while thinking on this and I've come to the conclusion that NFC cards don't really fit into HWI's expected workflow. With USB devices,there is a persistent connection between the device and the computer by virtue of having to plug it in. However with NFC, the connection is really ephemeral with the user expecting to be prompted to tap the card to perform an action. Using a NFC card with HWI would require having to but the card on a reader as if it were plugged in, but this is not how NFC is generally used. HWI is often used in a way where the type of device being used is unknown at the time of command invocation. Users either call `enumerate` to figure that out before issuing the command, or they use the fingerprint and let HWI do it automatically. Both of these require that the device is connected prior to the command being invoked. With NFC cards, this would require the caller to prompt the user to place their card on the reader and to not move it for some time. I don't think users expect to need to do that, nor for callers to know the type of device in order to make such a prompt.

Additionally, I question whether NFC cards would be used with HWI. HWI is a software targeted at desktop and laptop computers. Unless the user purchases a NFC reader, desktops and laptops are unlikely to be able to even interface with the card. ISTM the vast majority of users will only be able to use NFC from their phones, but HWI isn't expected to work on those devices. So I'm not sure that including Tapsigner support is worth the additional maintenance burden it will incur.

tl;dr I don't think Tapsigner is a good fit for HWI.

**Comment by nvk** [comment]
> > > Check coinkite store
> 
> > "IMPORTANT: Incompatible with Mk4 NFC features! Works only with NFC cards."
> 
> Wow, so Coldcard and Tapsigner use incompatible NFC standards?

Yes, was unavoidable. The hardware capabilities and the software stack are too far apart.
